lib/ansible/modules/web_infrastructure/ansible_tower/tower_credential.py

```diff
@@ -55,7 +55,12 @@
         - The ssh choice refers to a Tower Machine credential.
       required: True
       type: str
-      choices: ["ssh", "vault", "net", "scm", "aws", "vmware", "satellite6", "cloudforms", "gce", "azure_rm", "openstack", "rhv", "insights", "tower"]
+      choices: ["ssh", "vault", "net", "scm", "aws", "vmware", "satellite6", "cloudforms", "gce", "azure_rm", "openstack", "rhv", "insights", "tower", "cloud"]
+    credential_type:
+      description:
+        - Non built-in custom credential type name.
```

Comment:
built-in types also have a credential type name. Treating credential_type as a special case for custom credentials doesn't jive great with the design goals.

I wonder if we could make _either_ "kind" or "credential_type" required (irrelevant of custom vs. built-in types), and manage to get by that way. @ryanpetrello do you have any thoughts?

Reply:
Thanks for reviewing the request. Seems like support for legacy v1 Credential model, when only the "kind" field was used to identify credential type (from a static list of types) complicates the design. The latest Credential implementation references credential type record (uniquely identified by _credential type name_, _kind_ and _organization_). Seems like _kind_ is required across both versions.          

Depending on tower-cli credential type fetching logic, maybe an option to let tower-cli handle this complexity and just pass it optional _credential type name_, if provided. This way module would not need to handle differences between versions and built-in vs custom types.

Reply:
> The latest Credential implementation references credential type record (uniquely identified by credential type name, kind and organization). Seems like kind is required across both versions.

The v1 "kind" field is not the same as the v2 "kind" field. In v1, "kind" is a field on credentials and indicates something that corresponds to the v2 credential type (like aws, vmware, etc.). In v2, "kind" is a field on credential types, and only has a small handful of choices (like ssh, cloud, vault).

> Seems like support for legacy v1 Credential model, when only the "kind" field was used to identify credential type (from a static list of types) complicates the design.

Certainly agree that the issue complicates the design. Removal of v1 is near-term in AWX:

https://github.com/ansible/awx/pull/3413

> Depending on tower-cli credential type fetching logic, maybe an option to let tower-cli handle this complexity and just pass it optional credential type name, if provided.

Based on the prior points, this looks like it could involve abusing the "kind" field, and doing this would be very confusing because of those two different functions. I'm sorry I don't have answers here.